spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala

```diff
@@ -288,17 +307,92 @@ trait SnapshotManagement { self: DeltaLog =>
       }
       val lastCommitTimestamp = deltas.last.getModificationTime
 
-      validateDeltaVersions(deltasAfterCheckpoint, newCheckpointVersion, versionToLoad)
+      val deltasAndCompactedDeltasForLogSegment = useCompactedDeltasForLogSegment(
+        deltasAndCompactedDeltas,
+        deltasAfterCheckpoint,
+        latestCommitVersion = newVersion,
+        checkpointVersionToUse = newCheckpointVersion)
+
+      validateDeltaVersions(
+        deltasAndCompactedDeltasForLogSegment, newCheckpointVersion, versionToLoad)
 
       Some(LogSegment(
         logPath,
         newVersion,
-        deltasAfterCheckpoint,
+        deltasAndCompactedDeltasForLogSegment,
         checkpointProviderOpt,
         lastCommitTimestamp))
     }
   }
 
+  /**
+   * @param deltasAndCompactedDeltas - all deltas or compacted deltas which could be used
+   * @param deltasAfterCheckpoint - deltas after the last checkpoint file
+   * @param latestCommitVersion - commit version for which we are trying to create Snapshot for
+   * @param checkpointVersionToUse - underlying checkpoint version to use in Snapshot, -1 if no
+   *                               checkpoint is used.
+   * @return Returns a list of deltas/compacted-deltas which can be used to construct the
+   *         [[LogSegment]] instead of `deltasAfterCheckpoint`.
+   */
+  protected def useCompactedDeltasForLogSegment(
+      deltasAndCompactedDeltas: Seq[FileStatus],
+      deltasAfterCheckpoint: Array[FileStatus],
+      latestCommitVersion: Long,
+      checkpointVersionToUse: Long): Array[FileStatus] = {
+
+    val selectedDeltas = mutable.ArrayBuffer.empty[FileStatus]
+    var highestVersionSeen = checkpointVersionToUse
+    val commitRangeCovered = mutable.ArrayBuffer.empty[Long]
+    // track if there is at least 1 compacted delta in `deltasAndCompactedDeltas`
+    var hasCompactedDeltas = false
+    for (file <- deltasAndCompactedDeltas) {
+      val (startVersion, endVersion) = file match {
+        case CompactedDeltaFile(_, startVersion, endVersion) =>
+          hasCompactedDeltas = true
+          (startVersion, endVersion)
+        case DeltaFile(_, version) =>
+          (version, version)
+      }
+
+      if (highestVersionSeen < startVersion && endVersion <= latestCommitVersion) {
```

Comment:
This ensures we don't pick up overlapping compacted Deltas, right?
* First compaction can't straddle the starting version
* Neighboring compactions can't overlap
* Last compaction can't straddle latest version

We get a lot of work out of that one line, maybe worth a code comment?